File: scr/P122.f90
program main
    implicit none
    integer :: M, result
    double precision :: start_time, end_time, elapsed_time

    M = 200
    call cpu_time(start_time)
    result = P122(M)
    call cpu_time(end_time)
    elapsed_time = end_time - start_time
    print *, 'Resultado: ', result, ' (execution time ', elapsed_time, ' seconds)'
contains

    recursive function sum_chain(L, L_size, N, suma_actual, lenth_min, length_actual) result(res)
        integer, intent(in) :: L_size, N, suma_actual, length_actual
        integer :: lenth_min
        integer, intent(in) :: L(L_size)
        integer :: res, local_l, suma_actual_l, newL_size, i
        logical :: exists
        integer, allocatable :: newL(:)

        res = lenth_min
        do i = L_size, 1, -1
            local_l = L(i)
            suma_actual_l = suma_actual + local_l

            if (suma_actual_l > N) cycle
            if (suma_actual_l == N .and. length_actual < lenth_min) then
                res = length_actual
                return
            end if
            if (suma_actual_l < N .and. length_actual < lenth_min) then
                exists = any(L == suma_actual_l)
                if (.not. exists) then
                    newL_size = L_size + 1
                else
                    newL_size = L_size
                end if
                allocate(newL(newL_size))
                newL(1:L_size) = L
                if (.not. exists) newL(L_size+1) = suma_actual_l
                res = sum_chain(newL, newL_size, N, suma_actual_l, lenth_min, length_actual+1)
                deallocate(newL)
                if (res < lenth_min) lenth_min = res
            end if
        end do
    end function sum_chain

    function P122(M) result(total)
        integer, intent(in) :: M
        integer :: total, n, L(1)
        L(1) = 1
        total = 0
        do n = 1, M
            total = total + sum_chain(L, 1, n, 0, 9999, 0)
        end do
    end function P122

end program main

File: scr/P122.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>
/*  g++ -o P122 P122.cpp */ 
int sum_chain(int* L, int L_size, int N, int suma_actual, int lenth_min, int length_actual) {
    for (int i = L_size - 1; i >= 0; --i) {
        int l = L[i];
        int suma_actual_l = suma_actual + l;

        if (suma_actual_l > N) {
            continue;
        } else if (suma_actual_l == N && length_actual < lenth_min) {
            lenth_min = length_actual;
        } else if (suma_actual_l < N && length_actual < lenth_min) {
            int newL_size = L_size;
            bool exists = false;
            for (int j = 0; j < L_size; j++) {
                if (L[j] == suma_actual_l) {
                    exists = true;
                    break;
                }
            }
            
            if (!exists) {
                newL_size++;
            }

            int* newL = (int*) malloc(newL_size * sizeof(int));
            for (int j = 0; j < L_size; j++) {
                newL[j] = L[j];
            }

            if (!exists) {
                newL[L_size] = suma_actual_l;
            }

            lenth_min = sum_chain(newL, newL_size, N, suma_actual_l, lenth_min, length_actual + 1);
            free(newL);
        }
    }
    return lenth_min;
}

int P122(int M) {
    int total = 0;
    for (int n = 1; n <= M; ++n) {
        int L[1] = {1};
        total += sum_chain(L, 1, n, 0, 9999, 0);
    }
    return total;
}

int main() {
    clock_t start, end;
    double cpu_time_used;

    start = clock();
    int result = P122(200);
    end = clock();
    
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("Resultado: %d (execution time %f seconds)\n", result, cpu_time_used);
    
    return 0;
}
